src/core/io/src/4C_io_input_file.cpp

```diff
@@ -784,31 +784,106 @@ namespace Core::IO
+    // Dat file format
     if (pimpl_->content_by_section_.at(section_name).content.index() == 0)
     {
-      std::stringstream flattened_dat;
-      for (const auto& line : in_section(section_name))
-      {
-        std::string line_str(line.get_as_dat_style_string());
-        // Split the line into key-value according to the dat file format. Quote keys and values
-        // so the parser can identify them.
-        auto [key, value] = read_key_value(line_str);
-        flattened_dat << std::quoted(key) << " " << std::quoted(value) << " ";
+      // Dat format has too little structure and the interpretation of line breaks differs
+      // depending on the expected spec.
+      const auto* list_spec = dynamic_cast<const Internal::InputSpecTypeErasedImplementation<
+          InputSpecBuilders::Internal::ListSpec>*>(&spec.impl());
+
+      if (list_spec)
+      {
+        std::vector<InputParameterContainer> list_entries;
+        for (const auto& line : in_section(section_name))
+        {
+          Core::IO::ValueParser parser{line.get_as_dat_style_string(),
+              {.user_scope_message =
+                      "While parsing list entries in section '" + section_name + "': ",
+                  .base_path =
+                      std::filesystem::path(pimpl_->content_by_section_.at(section_name).file)
+                          .parent_path()}};
+          try
+          {
+            list_spec->wrapped.spec.fully_parse(parser, list_entries.emplace_back());
+          }
+          catch (const std::exception& e)
+          {
+            FOUR_C_THROW("Error while parsing list entries in section '%s': %s",
+                section_name.c_str(), e.what());
+          }
+        }
+        container.add_list(section_name, std::move(list_entries));
+      }
+      else
+      {
+        // Create a group in the dat file format by starting with the section name.
+        std::stringstream flattened_dat;
+        flattened_dat << std::quoted(section_name) << " ";
+        for (const auto& line : in_section(section_name))
+        {
+          std::string line_str(line.get_as_dat_style_string());
+          // Split the line into key-value according to the dat file format. Quote keys and values
+          // so the parser can identify them.
+          auto [key, value] = read_key_value(line_str);
+          flattened_dat << std::quoted(key) << " " << std::quoted(value) << " ";
+        }
+        std::string flattened_string = flattened_dat.str();
+        Core::IO::ValueParser parser{flattened_string,
+            {.base_path = std::filesystem::path(pimpl_->content_by_section_.at(section_name).file)
+                    .parent_path(),
+                .token_delimiter = '"'}};
+        try
+        {
+          spec.fully_parse(parser, container);
+        }
+        catch (const std::exception& e)
+        {
+          FOUR_C_THROW("Error while parsing section '%s': %s", section_name.c_str(), e.what());
+        }
```

Comment:
@maxfirmbach This restores better error messages for dat input.

Reply:
@sebproell Nice, thank you!